src/lib/core/diag.c

```diff
@@ -158,17 +168,45 @@ error_format_msg(struct error *e, const char *format, ...)
 void
 error_append_msg(struct error *e, const char *format, ...)
 {
-	va_list ap;
-	va_start(ap, format);
-	int prefix_len = strlen(e->errmsg);
-	char *msg = e->errmsg + prefix_len;
-	vsnprintf(msg, sizeof(e->errmsg) - prefix_len, format, ap);
-	va_end(ap);
+	va_list ap1, ap2;
+	va_start(ap1, format);
+	va_copy(ap2, ap1);
+	int static_buf_size = sizeof(e->errmsg_buf);
+	int curr_len = strlen(e->errmsg);
+	int len = vsnprintf(NULL, 0, format, ap1);
+	int new_size = curr_len + len + 1;
+	va_end(ap1);
+
+	if (new_size > static_buf_size) {
+		char *new_buf = xmalloc(new_size);
+		memcpy(new_buf, e->errmsg, curr_len + 1);
```

Comment:
Nit: why `curr_len + 1`?

Reply:
I feel a little uncomfortable leaving non-terminated strings in memory, even for a while.